File: siglume_api_sdk/diff.py
from __future__ import annotations

from dataclasses import asdict, dataclass, is_dataclass
from enum import Enum
from typing import Any, Mapping


class ChangeLevel(str, Enum):
    BREAKING = "breaking"
    WARNING = "warning"
    INFO = "info"


BreakingChange = ChangeLevel.BREAKING


@dataclass(frozen=True)
class Change:
    level: ChangeLevel
    path: str
    old: Any
    new: Any
    message: str

    @property
    def is_breaking(self) -> bool:
        return self.level == ChangeLevel.BREAKING

    def to_dict(self) -> dict[str, Any]:
        return {
            "level": self.level.value,
            "path": self.path,
            "old": self.old,
            "new": self.new,
            "message": self.message,
            "is_breaking": self.is_breaking,
        }


_MANIFEST_PERMISSION_ORDER = {
    "read-only": 0,
    "recommendation": 0,
    "action": 1,
    "payment": 2,
}
_TOOL_MANUAL_PERMISSION_ORDER = {
    "read_only": 0,
    "action": 1,
    "payment": 2,
}
_SPECIAL_MANIFEST_KEYS = {
    "version",
    "name",
    "short_description",
    "permission_class",
    "price_model",
    "currency",
    "jurisdiction",
}
_SPECIAL_TOOL_MANUAL_KEYS = {
    "input_schema",
    "output_schema",
    "permission_class",
    "settlement_mode",
    "currency",
    "side_effect_summary",
    "jurisdiction",
    "trigger_conditions",
    "do_not_use_when",
    "approval_summary_template",
}


def diff_manifest(*, old: Any, new: Any) -> list[Change]:
    old_payload = _normalize_manifest(old)
    new_payload = _normalize_manifest(new)
    changes: list[Change] = []
    emitted_keys: set[str] = set()

    _append_permission_class_change(
        changes,
        emitted_keys,
        key="permission_class",
        old_value=old_payload.get("permission_class"),
        new_value=new_payload.get("permission_class"),
        rank_map=_MANIFEST_PERMISSION_ORDER,
        escalate_message="Manifest permission_class escalated; existing callers may now require stronger approval.",
        downgrade_message="Manifest permission_class downgraded.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.BREAKING,
        key="price_model",
        old_value=old_payload.get("price_model"),
        new_value=new_payload.get("price_model"),
        message="Manifest price_model changed; billing compatibility may break existing installs.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.BREAKING,
        key="currency",
        old_value=old_payload.get("currency"),
        new_value=new_payload.get("currency"),
        message="Manifest currency changed.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.BREAKING,
        key="jurisdiction",
        old_value=old_payload.get("jurisdiction"),
        new_value=new_payload.get("jurisdiction"),
        message="Manifest jurisdiction changed.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.INFO,
        key="version",
        old_value=old_payload.get("version"),
        new_value=new_payload.get("version"),
        message="Manifest version changed.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.INFO,
        key="name",
        old_value=old_payload.get("name"),
        new_value=new_payload.get("name"),
        message="Manifest display name changed.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.INFO,
        key="short_description",
        old_value=old_payload.get("short_description"),
        new_value=new_payload.get("short_description"),
        message="Manifest short_description changed.",
    )

    for key in sorted(set(old_payload) | set(new_payload)):
        if key in emitted_keys or key in _SPECIAL_MANIFEST_KEYS:
            continue
        old_value = old_payload.get(key)
        new_value = new_payload.get(key)
        if _values_differ(old_value, new_value):
            changes.append(
                Change(
                    level=ChangeLevel.INFO,
                    path=key,
                    old=old_value,
                    new=new_value,
                    message=f"Manifest field '{key}' changed.",
                )
            )

    return _sort_changes(changes)


def diff_tool_manual(*, old: Any, new: Any) -> list[Change]:
    old_payload = _normalize_tool_manual(old)
    new_payload = _normalize_tool_manual(new)
    changes: list[Change] = []
    emitted_keys: set[str] = set()

    old_required = _string_list(_mapping(old_payload.get("input_schema")).get("required"))
    new_required = _string_list(_mapping(new_payload.get("input_schema")).get("required"))
    added_required = sorted(set(new_required) - set(old_required))
    removed_required = sorted(set(old_required) - set(new_required))
    if added_required:
        changes.append(
            Change(
                level=ChangeLevel.BREAKING,
                path="input_schema.required",
                old=old_required,
                new=new_required,
                message=f"input_schema.required added new required fields: {', '.join(added_required)}.",
            )
        )
        emitted_keys.add("input_schema")
    if removed_required:
        changes.append(
            Change(
                level=ChangeLevel.INFO,
                path="input_schema.required",
                old=old_required,
                new=new_required,
                message=f"input_schema.required removed fields: {', '.join(removed_required)}.",
            )
        )
        emitted_keys.add("input_schema")

    _append_permission_class_change(
        changes,
        emitted_keys,
        key="permission_class",
        old_value=old_payload.get("permission_class"),
        new_value=new_payload.get("permission_class"),
        rank_map=_TOOL_MANUAL_PERMISSION_ORDER,
        escalate_message="ToolManual permission_class escalated; existing callers may now require stronger approval.",
        downgrade_message="ToolManual permission_class downgraded.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.BREAKING,
        key="settlement_mode",
        old_value=old_payload.get("settlement_mode"),
        new_value=new_payload.get("settlement_mode"),
        message="ToolManual settlement_mode changed.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.BREAKING,
        key="currency",
        old_value=old_payload.get("currency"),
        new_value=new_payload.get("currency"),
        message="ToolManual currency changed.",
    )
    if _values_differ(old_payload.get("side_effect_summary"), new_payload.get("side_effect_summary")):
        old_side_effect = old_payload.get("side_effect_summary")
        new_side_effect = new_payload.get("side_effect_summary")
        if _is_blank(old_side_effect) and not _is_blank(new_side_effect):
            level = ChangeLevel.BREAKING
            message = "ToolManual side_effect_summary was added, introducing a new side-effect contract."
        elif not _is_blank(old_side_effect) and _is_blank(new_side_effect):
            level = ChangeLevel.INFO
            message = "ToolManual side_effect_summary was removed."
        else:
            level = ChangeLevel.BREAKING
            message = "ToolManual side_effect_summary changed."
        changes.append(
            Change(
                level=level,
                path="side_effect_summary",
                old=old_side_effect,
                new=new_side_effect,
                message=message,
            )
        )
        emitted_keys.add("side_effect_summary")
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.BREAKING,
        key="jurisdiction",
        old_value=old_payload.get("jurisdiction"),
        new_value=new_payload.get("jurisdiction"),
        message="ToolManual jurisdiction changed.",
    )

    old_output_fields = sorted(_mapping(_mapping(old_payload.get("output_schema")).get("properties")).keys())
    new_output_fields = sorted(_mapping(_mapping(new_payload.get("output_schema")).get("properties")).keys())
    added_output_fields = sorted(set(new_output_fields) - set(old_output_fields))
    removed_output_fields = sorted(set(old_output_fields) - set(new_output_fields))
    changed_output_fields = sorted(
        field_name
        for field_name in set(old_output_fields) & set(new_output_fields)
        if _values_differ(
            _mapping(_mapping(old_payload.get("output_schema")).get("properties")).get(field_name),
            _mapping(_mapping(new_payload.get("output_schema")).get("properties")).get(field_name),
        )
    )
    if added_output_fields:
        changes.append(
            Change(
                level=ChangeLevel.WARNING,
                path="output_schema.properties",
                old=old_output_fields,
                new=new_output_fields,
                message=f"output_schema added fields: {', '.join(added_output_fields)}.",
            )
        )
        emitted_keys.add("output_schema")
    if removed_output_fields:
        changes.append(
            Change(
                level=ChangeLevel.BREAKING,
                path="output_schema.properties",
                old=old_output_fields,
                new=new_output_fields,
                message=f"output_schema removed fields: {', '.join(removed_output_fields)}.",
            )
        )
        emitted_keys.add("output_schema")
    if changed_output_fields:
        changes.append(
            Change(
                level=ChangeLevel.WARNING,
                path="output_schema.properties",
                old=old_output_fields,
                new=new_output_fields,
                message=f"output_schema changed field definitions: {', '.join(changed_output_fields)}.",
            )
        )
        emitted_keys.add("output_schema")

    _append_large_list_change(
        changes,
        emitted_keys,
        key="trigger_conditions",
        old_items=_string_list(old_payload.get("trigger_conditions")),
        new_items=_string_list(new_payload.get("trigger_conditions")),
        warning_message="trigger_conditions changed substantially.",
        info_message="trigger_conditions changed.",
    )
    _append_large_list_change(
        changes,
        emitted_keys,
        key="do_not_use_when",
        old_items=_string_list(old_payload.get("do_not_use_when")),
        new_items=_string_list(new_payload.get("do_not_use_when")),
        warning_message="do_not_use_when changed substantially.",
        info_message="do_not_use_when changed.",
    )
    _append_value_change(
        changes,
        emitted_keys,
        level=ChangeLevel.WARNING,
        key="approval_summary_template",
        old_value=old_payload.get("approval_summary_template"),
        new_value=new_payload.get("approval_summary_template"),
        message="approval_summary_template changed.",
    )

    for key in sorted(set(old_payload) | set(new_payload)):
        if key in emitted_keys or key in _SPECIAL_TOOL_MANUAL_KEYS:
            continue
        old_value = old_payload.get(key)
        new_value = new_payload.get(key)
        if _values_differ(old_value, new_value):
            changes.append(
                Change(
                    level=ChangeLevel.INFO,
                    path=key,
                    old=old_value,
                    new=new_value,
                    message=f"ToolManual field '{key}' changed.",
                )
            )

    return _sort_changes(changes)


def _append_large_list_change(
    changes: list[Change],
    emitted_keys: set[str],
    *,
    key: str,
    old_items: list[str],
    new_items: list[str],
    warning_message: str,
    info_message: str,
) -> None:
    if old_items == new_items:
        return
    if _is_major_list_change(old_items, new_items):
        level = ChangeLevel.WARNING
        message = warning_message
    else:
        level = ChangeLevel.INFO
        message = info_message
    changes.append(Change(level=level, path=key, old=old_items, new=new_items, message=message))
    emitted_keys.add(key)


def _append_permission_class_change(
    changes: list[Change],
    emitted_keys: set[str],
    *,
    key: str,
    old_value: Any,
    new_value: Any,
    rank_map: Mapping[str, int],
    escalate_message: str,
    downgrade_message: str,
) -> None:
    if not _values_differ(old_value, new_value):
        return
    old_rank = rank_map.get(str(old_value)) if old_value is not None else None
    new_rank = rank_map.get(str(new_value)) if new_value is not None else None
    if old_rank is not None and new_rank is not None and new_rank > old_rank:
        level = ChangeLevel.BREAKING
        message = escalate_message
    elif old_rank is not None and new_rank is not None and new_rank < old_rank:
        level = ChangeLevel.INFO
        message = downgrade_message
    else:
        level = ChangeLevel.INFO
        message = f"{key} changed."
    changes.append(Change(level=level, path=key, old=old_value, new=new_value, message=message))
    emitted_keys.add(key)


def _append_value_change(
    changes: list[Change],
    emitted_keys: set[str],
    *,
    level: ChangeLevel,
    key: str,
    old_value: Any,
    new_value: Any,
    message: str,
) -> None:
    if _values_differ(old_value, new_value):
        changes.append(Change(level=level, path=key, old=old_value, new=new_value, message=message))
        emitted_keys.add(key)


def _normalize_manifest(value: Any) -> dict[str, Any]:
    payload = _mapping(_normalize_value(value))
    normalized = dict(payload)
    normalized["price_model"] = normalized.get("price_model") or "free"
    normalized["currency"] = normalized.get("currency") or "USD"
    # AppManifest defaults permission_class to "read-only" (hyphen form,
    # PermissionClass.READ_ONLY). Without this default, a legacy/minimal
    # manifest without permission_class compared against an upgraded
    # manifest would leave oldRank undefined and appendPermissionClassChange
    # would downgrade the permission escalation from BREAKING to INFO —
    # letting the diff CLI exit 0 on a genuinely breaking change.
    normalized["permission_class"] = normalized.get("permission_class") or "read-only"
    return normalized


def _normalize_tool_manual(value: Any) -> dict[str, Any]:
    payload = _mapping(_normalize_value(value))
    normalized = dict(payload)
    normalized["permission_class"] = normalized.get("permission_class") or "read_only"
    normalized["requires_connected_accounts"] = _string_list(normalized.get("requires_connected_accounts"))
    return normalized


def _normalize_value(value: Any) -> Any:
    if isinstance(value, Enum):
        return value.value
    if is_dataclass(value):
        return {key: _normalize_value(item) for key, item in asdict(value).items()}
    to_dict = getattr(value, "to_dict", None)
    if callable(to_dict):
        return _normalize_value(to_dict())
    if isinstance(value, Mapping):
        return {str(key): _normalize_value(item) for key, item in value.items()}
    if isinstance(value, (list, tuple, set)):
        return [_normalize_value(item) for item in value]
    return value


def _mapping(value: Any) -> dict[str, Any]:
    return dict(value) if isinstance(value, Mapping) else {}


def _string_list(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [str(item) for item in value if isinstance(item, str)]


def _values_differ(left: Any, right: Any) -> bool:
    return _stable_value(_normalize_value(left)) != _stable_value(_normalize_value(right))


def _is_blank(value: Any) -> bool:
    return not isinstance(value, str) or value.strip() == ""


def _is_major_list_change(old_items: list[str], new_items: list[str]) -> bool:
    old_set = {item.strip().lower() for item in old_items if item.strip()}
    new_set = {item.strip().lower() for item in new_items if item.strip()}
    if old_set == new_set:
        return False
    if not old_set or not new_set:
        return True
    union = old_set | new_set
    if not union:
        return False
    similarity = len(old_set & new_set) / len(union)
    return similarity < 0.5 or (similarity == 0.5 and abs(len(old_set) - len(new_set)) >= 1)


def _sort_changes(changes: list[Change]) -> list[Change]:
    priority = {
        ChangeLevel.BREAKING: 0,
        ChangeLevel.WARNING: 1,
        ChangeLevel.INFO: 2,
    }
    return sorted(changes, key=lambda item: (priority[item.level], item.path))


def _stable_value(value: Any) -> Any:
    if isinstance(value, Mapping):
        return {
            key: _stable_value(value[key])
            for key in sorted(value)
        }
    if isinstance(value, list):
        return [_stable_value(item) for item in value]
    return value


__all__ = [
    "BreakingChange",
    "Change",
    "ChangeLevel",
    "diff_manifest",
    "diff_tool_manual",
]

File: siglume-api-sdk-ts/src/diff.ts
import type { AppManifest, JsonValue, ToolManual } from "./types";
import { coerceMapping, isRecord, toJsonable } from "./utils";

const MANIFEST_PERMISSION_ORDER: Record<string, number> = {
  "read-only": 0,
  recommendation: 0,
  action: 1,
  payment: 2,
};
const TOOL_MANUAL_PERMISSION_ORDER: Record<string, number> = {
  read_only: 0,
  action: 1,
  payment: 2,
};
const SPECIAL_MANIFEST_KEYS = new Set([
  "version",
  "name",
  "short_description",
  "permission_class",
  "price_model",
  "currency",
  "jurisdiction",
]);
const SPECIAL_TOOL_MANUAL_KEYS = new Set([
  "input_schema",
  "output_schema",
  "permission_class",
  "settlement_mode",
  "currency",
  "side_effect_summary",
  "jurisdiction",
  "trigger_conditions",
  "do_not_use_when",
  "approval_summary_template",
]);

export const ChangeLevel = {
  BREAKING: "breaking",
  WARNING: "warning",
  INFO: "info",
} as const;
export type ChangeLevel = (typeof ChangeLevel)[keyof typeof ChangeLevel];

export const BreakingChange = ChangeLevel.BREAKING;

export interface Change {
  level: ChangeLevel;
  path: string;
  old: JsonValue | Record<string, unknown> | unknown[] | null;
  new: JsonValue | Record<string, unknown> | unknown[] | null;
  message: string;
  is_breaking: boolean;
}

export function diff_manifest(options: {
  old: AppManifest | Record<string, unknown>;
  new: AppManifest | Record<string, unknown>;
}): Change[] {
  const oldPayload = normalizeManifest(options.old);
  const newPayload = normalizeManifest(options.new);
  const changes: Change[] = [];
  const emittedKeys = new Set<string>();

  appendPermissionClassChange(
    changes,
    emittedKeys,
    "permission_class",
    oldPayload.permission_class,
    newPayload.permission_class,
    MANIFEST_PERMISSION_ORDER,
    "Manifest permission_class escalated; existing callers may now require stronger approval.",
    "Manifest permission_class downgraded.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.BREAKING,
    "price_model",
    oldPayload.price_model,
    newPayload.price_model,
    "Manifest price_model changed; billing compatibility may break existing installs.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.BREAKING,
    "currency",
    oldPayload.currency,
    newPayload.currency,
    "Manifest currency changed.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.BREAKING,
    "jurisdiction",
    oldPayload.jurisdiction,
    newPayload.jurisdiction,
    "Manifest jurisdiction changed.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.INFO,
    "version",
    oldPayload.version,
    newPayload.version,
    "Manifest version changed.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.INFO,
    "name",
    oldPayload.name,
    newPayload.name,
    "Manifest display name changed.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.INFO,
    "short_description",
    oldPayload.short_description,
    newPayload.short_description,
    "Manifest short_description changed.",
  );

  for (const key of [...new Set([...Object.keys(oldPayload), ...Object.keys(newPayload)])].sort()) {
    if (emittedKeys.has(key) || SPECIAL_MANIFEST_KEYS.has(key)) {
      continue;
    }
    const oldValue = oldPayload[key];
    const newValue = newPayload[key];
    if (valuesDiffer(oldValue, newValue)) {
      changes.push(createChange(ChangeLevel.INFO, key, oldValue, newValue, `Manifest field '${key}' changed.`));
    }
  }

  return sortChanges(changes);
}

export function diff_tool_manual(options: {
  old: ToolManual | Record<string, unknown>;
  new: ToolManual | Record<string, unknown>;
}): Change[] {
  const oldPayload = normalizeToolManual(options.old);
  const newPayload = normalizeToolManual(options.new);
  const changes: Change[] = [];
  const emittedKeys = new Set<string>();

  const oldRequired = stringList(toRecordValue(oldPayload.input_schema).required);
  const newRequired = stringList(toRecordValue(newPayload.input_schema).required);
  const addedRequired = [...new Set(newRequired.filter((fieldName) => !oldRequired.includes(fieldName)))].sort();
  const removedRequired = [...new Set(oldRequired.filter((fieldName) => !newRequired.includes(fieldName)))].sort();
  if (addedRequired.length > 0) {
    changes.push(
      createChange(
        ChangeLevel.BREAKING,
        "input_schema.required",
        oldRequired,
        newRequired,
        `input_schema.required added new required fields: ${addedRequired.join(", ")}.`,
      ),
    );
    emittedKeys.add("input_schema");
  }
  if (removedRequired.length > 0) {
    changes.push(
      createChange(
        ChangeLevel.INFO,
        "input_schema.required",
        oldRequired,
        newRequired,
        `input_schema.required removed fields: ${removedRequired.join(", ")}.`,
      ),
    );
    emittedKeys.add("input_schema");
  }

  appendPermissionClassChange(
    changes,
    emittedKeys,
    "permission_class",
    oldPayload.permission_class,
    newPayload.permission_class,
    TOOL_MANUAL_PERMISSION_ORDER,
    "ToolManual permission_class escalated; existing callers may now require stronger approval.",
    "ToolManual permission_class downgraded.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.BREAKING,
    "settlement_mode",
    oldPayload.settlement_mode,
    newPayload.settlement_mode,
    "ToolManual settlement_mode changed.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.BREAKING,
    "currency",
    oldPayload.currency,
    newPayload.currency,
    "ToolManual currency changed.",
  );
  if (valuesDiffer(oldPayload.side_effect_summary, newPayload.side_effect_summary)) {
    const oldSideEffect = oldPayload.side_effect_summary ?? null;
    const newSideEffect = newPayload.side_effect_summary ?? null;
    const level =
      isBlank(oldSideEffect) && !isBlank(newSideEffect)
        ? ChangeLevel.BREAKING
        : !isBlank(oldSideEffect) && isBlank(newSideEffect)
          ? ChangeLevel.INFO
          : ChangeLevel.BREAKING;
    const message =
      isBlank(oldSideEffect) && !isBlank(newSideEffect)
        ? "ToolManual side_effect_summary was added, introducing a new side-effect contract."
        : !isBlank(oldSideEffect) && isBlank(newSideEffect)
          ? "ToolManual side_effect_summary was removed."
          : "ToolManual side_effect_summary changed.";
    changes.push(
      createChange(
        level,
        "side_effect_summary",
        oldSideEffect,
        newSideEffect,
        message,
      ),
    );
    emittedKeys.add("side_effect_summary");
  }
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.BREAKING,
    "jurisdiction",
    oldPayload.jurisdiction,
    newPayload.jurisdiction,
    "ToolManual jurisdiction changed.",
  );

  const oldOutputProperties = toRecordValue(toRecordValue(oldPayload.output_schema).properties);
  const newOutputProperties = toRecordValue(toRecordValue(newPayload.output_schema).properties);
  const oldOutputFields = Object.keys(oldOutputProperties).sort();
  const newOutputFields = Object.keys(newOutputProperties).sort();
  const addedOutputFields = [...new Set(newOutputFields.filter((fieldName) => !oldOutputFields.includes(fieldName)))].sort();
  const removedOutputFields = [...new Set(oldOutputFields.filter((fieldName) => !newOutputFields.includes(fieldName)))].sort();
  const changedOutputFields = [...new Set(
    oldOutputFields
      .filter((fieldName) => newOutputFields.includes(fieldName))
      .filter((fieldName) => valuesDiffer(oldOutputProperties[fieldName], newOutputProperties[fieldName])),
  )].sort();
  if (addedOutputFields.length > 0) {
    changes.push(
      createChange(
        ChangeLevel.WARNING,
        "output_schema.properties",
        oldOutputFields,
        newOutputFields,
        `output_schema added fields: ${addedOutputFields.join(", ")}.`,
      ),
    );
    emittedKeys.add("output_schema");
  }
  if (removedOutputFields.length > 0) {
    changes.push(
      createChange(
        ChangeLevel.BREAKING,
        "output_schema.properties",
        oldOutputFields,
        newOutputFields,
        `output_schema removed fields: ${removedOutputFields.join(", ")}.`,
      ),
    );
    emittedKeys.add("output_schema");
  }
  if (changedOutputFields.length > 0) {
    changes.push(
      createChange(
        ChangeLevel.WARNING,
        "output_schema.properties",
        oldOutputFields,
        newOutputFields,
        `output_schema changed field definitions: ${changedOutputFields.join(", ")}.`,
      ),
    );
    emittedKeys.add("output_schema");
  }

  appendLargeListChange(
    changes,
    emittedKeys,
    "trigger_conditions",
    stringList(oldPayload.trigger_conditions),
    stringList(newPayload.trigger_conditions),
    "trigger_conditions changed substantially.",
    "trigger_conditions changed.",
  );
  appendLargeListChange(
    changes,
    emittedKeys,
    "do_not_use_when",
    stringList(oldPayload.do_not_use_when),
    stringList(newPayload.do_not_use_when),
    "do_not_use_when changed substantially.",
    "do_not_use_when changed.",
  );
  appendValueChange(
    changes,
    emittedKeys,
    ChangeLevel.WARNING,
    "approval_summary_template",
    oldPayload.approval_summary_template,
    newPayload.approval_summary_template,
    "approval_summary_template changed.",
  );

  for (const key of [...new Set([...Object.keys(oldPayload), ...Object.keys(newPayload)])].sort()) {
    if (emittedKeys.has(key) || SPECIAL_TOOL_MANUAL_KEYS.has(key)) {
      continue;
    }
    const oldValue = oldPayload[key];
    const newValue = newPayload[key];
    if (valuesDiffer(oldValue, newValue)) {
      changes.push(createChange(ChangeLevel.INFO, key, oldValue, newValue, `ToolManual field '${key}' changed.`));
    }
  }

  return sortChanges(changes);
}

function appendLargeListChange(
  changes: Change[],
  emittedKeys: Set<string>,
  key: string,
  oldItems: string[],
  newItems: string[],
  warningMessage: string,
  infoMessage: string,
): void {
  if (sameArray(oldItems, newItems)) {
    return;
  }
  const level = isMajorListChange(oldItems, newItems) ? ChangeLevel.WARNING : ChangeLevel.INFO;
  changes.push(createChange(level, key, oldItems, newItems, level === ChangeLevel.WARNING ? warningMessage : infoMessage));
  emittedKeys.add(key);
}

function appendPermissionClassChange(
  changes: Change[],
  emittedKeys: Set<string>,
  key: string,
  oldValue: unknown,
  newValue: unknown,
  rankMap: Record<string, number>,
  escalateMessage: string,
  downgradeMessage: string,
): void {
  if (!valuesDiffer(oldValue, newValue)) {
    return;
  }
  const oldRank = typeof oldValue === "string" ? rankMap[oldValue] : undefined;
  const newRank = typeof newValue === "string" ? rankMap[newValue] : undefined;
  const level =
    oldRank !== undefined && newRank !== undefined && newRank > oldRank
      ? ChangeLevel.BREAKING
      : ChangeLevel.INFO;
  const message =
    oldRank !== undefined && newRank !== undefined && newRank > oldRank
      ? escalateMessage
      : oldRank !== undefined && newRank !== undefined && newRank < oldRank
        ? downgradeMessage
        : `${key} changed.`;
  changes.push(createChange(level, key, oldValue, newValue, message));
  emittedKeys.add(key);
}

function appendValueChange(
  changes: Change[],
  emittedKeys: Set<string>,
  level: ChangeLevel,
  key: string,
  oldValue: unknown,
  newValue: unknown,
  message: string,
): void {
  if (valuesDiffer(oldValue, newValue)) {
    changes.push(createChange(level, key, oldValue, newValue, message));
    emittedKeys.add(key);
  }
}

function normalizeManifest(value: AppManifest | Record<string, unknown>): Record<string, unknown> {
  const payload = coerceMapping(value, "manifest");
  return {
    ...payload,
    price_model: payload.price_model ?? "free",
    currency: payload.currency ?? "USD",
    // AppManifest defaults permission_class to "read-only" (hyphen form,
    // PermissionClass.READ_ONLY). Without this default, a legacy / minimal
    // manifest without permission_class compared against an upgraded one
    // would leave oldRank undefined and appendPermissionClassChange would
    // downgrade the permission escalation from BREAKING to INFO — letting
    // the diff CLI exit 0 on a genuinely breaking change.
    permission_class: payload.permission_class ?? "read-only",
  };
}

function normalizeToolManual(value: ToolManual | Record<string, unknown>): Record<string, unknown> {
  const payload = coerceMapping(value, "tool manual");
  return {
    ...payload,
    permission_class: payload.permission_class ?? "read_only",
    requires_connected_accounts: stringList(payload.requires_connected_accounts),
  };
}

function createChange(
  level: ChangeLevel,
  path: string,
  oldValue: unknown,
  newValue: unknown,
  message: string,
): Change {
  return {
    level,
    path,
    old: normalizeForChange(oldValue),
    new: normalizeForChange(newValue),
    message,
    is_breaking: level === ChangeLevel.BREAKING,
  };
}

function normalizeForChange(value: unknown): JsonValue | Record<string, unknown> | unknown[] | null {
  if (value === undefined) {
    return null;
  }
  const normalized = toJsonable(value);
  if (normalized === undefined) {
    return null;
  }
  return normalized as JsonValue | Record<string, unknown> | unknown[] | null;
}

function toRecordValue(value: unknown): Record<string, unknown> {
  return isRecord(value) ? { ...value } : {};
}

function stringList(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value.filter((item): item is string => typeof item === "string");
}

function valuesDiffer(left: unknown, right: unknown): boolean {
  return JSON.stringify(stableValue(normalizeForChange(left))) !== JSON.stringify(stableValue(normalizeForChange(right)));
}

function sameArray(left: string[], right: string[]): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function isBlank(value: unknown): boolean {
  return typeof value !== "string" || value.trim().length === 0;
}

function isMajorListChange(oldItems: string[], newItems: string[]): boolean {
  const oldSet = new Set(oldItems.map((item) => item.trim().toLowerCase()).filter(Boolean));
  const newSet = new Set(newItems.map((item) => item.trim().toLowerCase()).filter(Boolean));
  if (oldSet.size === newSet.size && [...oldSet].every((item) => newSet.has(item))) {
    return false;
  }
  if (oldSet.size === 0 || newSet.size === 0) {
    return true;
  }
  const union = new Set([...oldSet, ...newSet]);
  const intersection = [...oldSet].filter((item) => newSet.has(item));
  const similarity = intersection.length / union.size;
  return similarity < 0.5 || (similarity === 0.5 && Math.abs(oldSet.size - newSet.size) >= 1);
}

function sortChanges(changes: Change[]): Change[] {
  const priority: Record<ChangeLevel, number> = {
    [ChangeLevel.BREAKING]: 0,
    [ChangeLevel.WARNING]: 1,
    [ChangeLevel.INFO]: 2,
  };
  return [...changes].sort((left, right) => {
    const levelDelta = priority[left.level] - priority[right.level];
    return levelDelta !== 0 ? levelDelta : left.path.localeCompare(right.path);
  });
}

function stableValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((item) => stableValue(item));
  }
  if (isRecord(value)) {
    return Object.fromEntries(
      Object.keys(value)
        .sort()
        .map((key) => [key, stableValue((value as Record<string, unknown>)[key])]),
    );
  }
  return value;
}
